pkg/controller/machineset/machineset_controller.go

```diff
@@ -301,6 +302,33 @@ func (r *ReconcileMachineSet) syncReplicas(ms *clusterv1.MachineSet, machines []
 				i+1, diff, *(ms.Spec.Replicas), len(machines))
 
 			machine := r.getNewMachine(ms)
+
+			// Clone and set the infrastructure reference.
+			infraConfig, err := external.Clone(r.Client, &machine.Spec.InfrastructureRef, machine.Namespace)
+			if err != nil {
+				return errors.Wrapf(err, "failed to clone infrastructure configuration for MachineSet %q in namespace %q", ms.Name, ms.Namespace)
+			}
+			machine.Spec.InfrastructureRef = corev1.ObjectReference{
+				APIVersion: infraConfig.GetAPIVersion(),
+				Kind:       infraConfig.GetKind(),
+				Namespace:  infraConfig.GetNamespace(),
+				Name:       infraConfig.GetName(),
+			}
+
+			// Clone and set the bootstrap reference, if any.
+			if machine.Spec.Bootstrap.ConfigRef != nil {
+				bootstrapConfig, err := external.Clone(r.Client, machine.Spec.Bootstrap.ConfigRef, machine.Namespace)
+				if err != nil {
+					return errors.Wrapf(err, "failed to clone bootstrap configuration for MachineSet %q in namespace %q", ms.Name, ms.Namespace)
+				}
+				machine.Spec.Bootstrap.ConfigRef = &corev1.ObjectReference{
+					APIVersion: bootstrapConfig.GetAPIVersion(),
+					Kind:       bootstrapConfig.GetKind(),
+					Namespace:  bootstrapConfig.GetNamespace(),
+					Name:       bootstrapConfig.GetName(),
+				}
+			}
+
 			if err := r.Client.Create(context.Background(), machine); err != nil {
```

Comment:
Shouldn't the created referenced objects be cleaned up if this `Create` fails? Otherwise we are potentially leaving orphaned objects behind.

Reply:
That's true, I'll add a defer

Reply:
Done!